src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs

```diff
@@ -126,164 +139,270 @@ public override long Position
+            return ReadDataAfterWaitingForWriter(buffer, offset, count);
+        }
+
+        /// <summary>
+        /// Common routine used by both Read and ReadAsync to read available data after waiting
+        /// for the writer to publish.
+        /// </summary>
+        private int ReadDataAfterWaitingForWriter(byte[] buffer, int offset, int count)
+        {
+            // We should now have data to read, as long as cancellation hasn't been requested.
+            if (!_completionSignal.IsCancellationRequested)
+            {
+                Debug.Assert(_writerPublishedDataEvent.CurrentCount == 0, "Since cancellation wasn't requested, expected semaphore to have 0 count");
+                lock (LockObject)
                 {
-                    // If both are set, unblock the reader based on data availability
-                    lock (LockObject)
+                    Debug.Assert(_completionSignal.IsCancellationRequested || _availableData != IntPtr.Zero, "Expected writer to publish non-null pointer");
+                    Debug.Assert(_completionSignal.IsCancellationRequested || _availableDataLength > 0, "Expected writer to publish positive amount of data");
+                    if (_availableDataLength > 0)
                     {
-                        return (_innerStream != null);
+                        return ConsumePublishedData(buffer, offset, count);
                     }
                 }
-                return false;
             }
+
+            // There's no more data to read due to the stream shutting down.  If we shut down due to a failure,
+            // propagate the exception.  Otherwise, return 0 to indicate that the stream is complete.
+            Debug.Assert(_completionSignal.IsCancellationRequested, "Cancellation must have been requested to get here");
+            ExceptionDispatchInfo edi = _storedException;
+            if (edi != null)
+            {
+                edi.Throw();
+            }
+            return 0;
         }
 
-        private void SignalWriter()
+        /// <summary>
+        /// Copies at most <paramref name="count"/> bytes from the published data into the 
+        /// supplied buffer at the specified offset.  The published data is then updated
+        /// accordingly to reflect the consumed data.
+        /// </summary>
+        /// <returns>The number of bytes copied.</returns>
+        private int ConsumePublishedData(byte[] buffer, int offset, int count)
         {
-            Debug.Assert(Monitor.IsEntered(LockObject), "lock wasn't held");
-            _writerHasDataEvent.Reset();
-            _readerRequestingDataEvent.Set();
+            Debug.Assert(Monitor.IsEntered(LockObject), "Published data must only be accessed while holding the lock");
+
+            // Copy the data into the read buffer
+            int bytesToRead = (int)Math.Min(_availableDataLength, count);
+            Marshal.Copy(_availableData, buffer, offset, bytesToRead);
```

Comment:
I was worried about the scenario where libcurl may have consumed the native pointer, that it gave us in the curl-callback. I now realize that wont happen. I was wrong.